aclk/aclk_query_queue.c

```diff
@@ -49,10 +49,36 @@ int aclk_queue_query(aclk_query_t query)
 {
     int ret = _aclk_queue_query(query);
     if (!ret) {
+        // local cache of query type before we wake up query thread, which may
+        // free the query in a race.
+        aclk_query_type_t qtype = query->type;
         QUERY_THREAD_WAKEUP;
+
         if (aclk_stats_enabled) {
+            // get target query type metric before lock so we keep lock for
+            // minimal time.
+            volatile uint32_t *metric = NULL;
```

Comment:
could we perhaps turn this into inline func and change to `switch` just for readability?

Reply:
Yes that may be useful in other places indeed.